lumicks/pylake/kymotracker/detail/velocity.py

```diff
@@ -0,0 +1,254 @@
+    return min(filtered_results if len(filtered_results) else results, key=lambda r: r.rss)
+
+
+def _optimize_linear(x, y):
+    """OLS linear regression of x and y.
+
+    Parameters
+    ----------
+    x : np.ndarray
+        Dependent variable (time).
+    y : np.ndarray
+        Independent variable (position).
+    """
+    n_coeffs = 2
+    n_samples = len(x)
+
+    design_matrix = np.vstack((np.ones(n_samples), x)).T
+    xtx = np.linalg.pinv(np.matmul(design_matrix.T, design_matrix))
+    coeffs = np.dot(
+        xtx,
+        np.dot(design_matrix.T, y),
+    )
+
+    fit = np.dot(design_matrix, coeffs)
+    residuals = fit - y
+    rss = np.sum(residuals**2)
+    bic = n_samples * np.log(rss / n_samples) + n_coeffs * np.log(n_samples)
+
+    intercept, slope = coeffs
+
+    cov = (rss / (n_samples - n_coeffs)) * xtx
+    intercept_std, slope_std = np.sqrt(np.diag(cov))
+
+    return PiecewiseModel(
+        intercept, [slope], [], intercept_std, [slope_std], [], rss, bic, "converged"
+    )
+
+
+def _optimize_breakpoints(x, y, n_breakpoints, tol, max_iter):
+    r"""
+    Optimization of breakpoint locations for a continuous piecewise linear model
+
+    Parameters
+    ----------
+    x : np.ndarray
+        Dependent variable (time).
+    y : np.ndarray
+        Independent variable (position).
+    n_breakpoints : int
+        Number of requested breakpoints.
+    tol : float
+        Convergence tolerance. Ignored if `n_breakpoints == 0`.
+    max_iter : int
+        Maximum number of iterations allowed in the optimization algorithm. Ignored if
+        `n_breakpoints == 0`.
+    n_restarts : int
+        Number of times to run the optimization algorithm with breakpoint initial guesses drawn
+        from a uniform distribution. Ignored if `n_breakpoints == 0`.
+    """
+    n_samples = len(x)
+    n_coeffs = 2 + 2 * n_breakpoints
+
+    def fit_iteration(breakpoints):
+        r"""
+        Single iteration of the breakpoint estimation algorithm, solving Eq 7. corresponding
+        to linearized form of piecewise continuous model, with nonlinear breakpoints term:
+
+        .. math::
+
+            \alpha x + \beta U + \gamma V
+
+        where :math:`\alpha` is the slope of the first section, :math:`\beta` models the
+        difference-in-slopes of the subsequent sections, and :math:`\gamma` is a reparameterization
+        of the breakpoints :math:`\psi`.
+
+        Given the current breakpoints guess :math:`\psi^{(0)}` and the ML estimates of above
+        parameters using OLS, the updated breakpoints can be calculated as:
+
+        .. math::
+
+            \psi = \frac{\gamma}{\beta} + \psi^{(0)}
+        """
+        diff_term = x - breakpoints[:, np.newaxis]
+        heavi = np.vstack([np.heaviside(d, -1) for d in diff_term])
+        u = diff_term * heavi
+        v = -heavi
+        design_matrix = np.vstack((np.ones(x.size), x, u, v)).T
+
+        xtx = np.linalg.pinv(np.matmul(design_matrix.T, design_matrix))
```

Comment:
There's a shorthand for matmul.
```suggestion
        xtx = np.linalg.pinv(design_matrix.T @ design_matrix)
```